usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -2180,16 +2043,62 @@ static struct parsefuncs parseupdatefuncs = {
 static CK_RV update_opencryptoki_conf(CK_SLOT_ID slot_id, char *location)
 {
     char dst_file[PATH_MAX], src_file[PATH_MAX], fname[PATH_MAX+20];
+    struct ConfigBaseNode *config = NULL, *c;
+    struct ConfigVersionValNode *v;
+    struct ConfigIdxStructNode *slot;
     FILE *fp_w = NULL;
     CK_RV ret;
     int rc;
-    struct parseupdate parseupdate;
 
     TRACE_INFO("Updating config file ...\n");
 
     /* Open current conf file for read */
     snprintf(src_file, PATH_MAX, "%s/%s", location, "opencryptoki.conf");
 
+    config = config_parse(src_file, TRUE);
+    if (config == NULL) {
+        TRACE_ERROR("failed to parse config file %s\n", src_file);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Update the tokeversion for the selected slot */
+    slot = confignode_findidx(config, "slot", slot_id);
+    if (slot == NULL) {
+        TRACE_ERROR("failed to find slot %lu in config file %s\n", slot_id,
+                    src_file);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    c = confignode_find(slot->value, "tokversion");
+    if (c != NULL) {
+        /* modify existing tokversion */
+        if (confignode_hastype(c, CT_VERSIONVAL)) {
+            confignode_to_versionval(c)->value = TOKVERSION_312;
+        } else if (confignode_hastype(c, CT_STRINGVAL)) {
+            free(confignode_to_stringval(c)->value);
+            confignode_to_stringval(c)->value = strdup(TOKVERSION_312_STRING);
```

Comment:
Maybe we should check the result of `strdup`?

Reply:
fixed